crates/p2p/src/swarm/dto/dht_record.rs

```diff
@@ -0,0 +1,116 @@
+//! Our DHT record type.
+
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use libp2p::{Multiaddr, PeerId, identity::PublicKey};
+use serde::{Deserialize, Deserializer, Serialize, de};
+
+use crate::{
+    signer::ApplicationSigner,
+    swarm::serializing::{
+        pubkey_serialization::pubkey_serializer, signature_serialization::signature_serializer,
+    },
+};
+
+/// Protocol version for DHT records.
+pub(crate) const DHT_PROTOCOL_VERSION: u8 = 0;
+
+/// Protocol identifier for DHT records.
+pub(crate) const DHT_PROTOCOL_ID: u8 = 0;
+
+/// Record structure for DHT.
+/// (De)serializable via serde.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RecordData {
+    /// Protocol version.
+    pub version: u8,
+    /// Protocol identifier.
+    pub protocol: u8,
+    /// The application public key (Ed25519).
+    #[serde(with = "pubkey_serializer")]
+    pub app_public_key: PublicKey,
+    /// Transport ID (PeerId).
+    pub transport_id: PeerId,
+    /// Multiaddresses.
+    pub multiaddresses: Vec<Multiaddr>,
+    /// Timestamp of message creation.
+    pub date: u64,
+}
+
+impl RecordData {
+    pub(crate) fn new(
+        app_public_key: PublicKey,
+        transport_id: PeerId,
+        multiaddresses: Vec<Multiaddr>,
+    ) -> Self {
+        let timestamp = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs();
+
+        Self {
+            version: DHT_PROTOCOL_VERSION,
+            protocol: DHT_PROTOCOL_ID,
+            app_public_key,
+            transport_id,
+            date: timestamp,
+            multiaddresses,
+        }
+    }
+}
+
+/// Signed DHT record.
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub struct SignedRecord {
```

Comment:
Let's have the same `SignedMessage` for all types paramed with some type `M`: Ser + Deser + PubKey